sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala

```diff
@@ -0,0 +1,456 @@
+
+  override val aggBufferAttributes: Seq[AttributeReference] = Seq()
+
+  override val inputAggBufferAttributes: Seq[AttributeReference] = Seq()
+
+  private var summary: QuantileSummaries = null
+
+  override def initialize(buffer: MutableRow): Unit = {
+    // Our `PercentileApprox` function takes a `B` parameter, but the underlying GK algorithm takes
+    // a `relativeError` parameter, so we need to convert `B` to `relativeError`.
+    // Please refer to SPARK-16283 for details.
+    val relativeError = PercentileApprox.bToRelativeError(B)
+    summary = new QuantileSummaries(QuantileSummaries.defaultCompressThreshold, relativeError)
+  }
+
+  override def update(buffer: MutableRow, input: InternalRow): Unit = {
+    val value = child.eval(input) match {
+      case o: Byte => o.toDouble
+      case o: Short => o.toDouble
+      case o: Int => o.toDouble
+      case o: Long => o.toDouble
+      case o: Float => o.toDouble
+      case o: Decimal => o.toDouble
+      case o: Double => o
+    }
+
+    summary.insert(value)
+  }
+
+  override def merge(buffer: MutableRow, inputBuffer: InternalRow): Unit = {
+    sys.error("PercentileApprox does not support partial aggregation.")
+  }
+
+  override def eval(buffer: InternalRow): Any = {
+    // summary must be compressed before being queried
+    summary = summary.compress()
+
+    if (resultAsArray) {
+      // return the result as an array of doubles, or return null for empty inputs
+      if (summary.count > 0) new GenericArrayData(percentiles.map { summary.query(_) }) else null
+    }
+    else {
+      // return the result as a double, or return null for empty inputs
+      if (summary.count > 0) summary.query(percentiles.head) else null
+    }
+  }
+
+  private def childrenSQL: String =
+    if (bExpr.isDefined) {
+      s"${child.sql}, ${percentilesExpr.toString}, ${bExpr.get.toString}"
+    }
+    else {
+      s"${child.sql}, ${percentilesExpr.toString}"
+    }
+
+  override def sql: String = s"$prettyName($childrenSQL)"
+
+  override def sql(isDistinct: Boolean): String = {
+    val distinct = if (isDistinct) "DISTINCT " else ""
+    s"$prettyName($distinct$childrenSQL)"
+  }
+}
+
+object PercentileApprox {
+
+  // Using 1000 as a default achieves a relativeError of 0.001
+  private[sql] val B_DEFAULT = 1000
+
+  // Our `PercentileApprox` function takes a `B` parameter, but the underlying GK algorithm takes
+  // a `relativeError` parameter, so we need to convert `B` to `relativeError`.
+  // Please refer to SPARK-16283 for details.
+  private[sql] def bToRelativeError(B: Int): Double = Math.max(1.0d / B, 0.001)
+
+  /**
+   * Validates the percentile(s) expression and extract the percentile(s).
+   * Returns the extracted percentile(s) and an indicator of whether it's an array.
+   */
+  private def validatePercentilesLiteral(exp: Expression): (Seq[Double], Boolean) = {
+    def withinRange(v: Double): Boolean = 0.0 <= v && v <= 1.0
+    exp match {
```

Comment:
You could also pattern match on the result of the `exp.eval()`. That would be way easier.